### PR TITLE
Support content management offerings, versions, catalogs

### PR DESCRIPTION
**Pull request by JonWoodlief**
You can run the acceptance tests with the following environment variables-

export CATMGMT_CLUSTERID="<cluster-id-roks-cluster>"
export CATMGMT_CLUSTERREGION="<region-of-roks-cluster>"
export TESTARGS="-run TestAccIBMCm"
export IBMCLOUD_IAM_API_ENDPOINT="https://iam.test.cloud.ibm.com"
export IBMCLOUD_CATALOG_MANAGEMENT_API_ENDPOINT="https://dev-cm.globalcatalog.test.cloud.ibm.com/api/v1-beta"

### CODE REVIEW

**Comment by JonWoodlief** [comment]
Screen Shot 2021-03-24 at 11 00 29 PM

**Comment by kavya498** [comment]
@JonWoodlief ,
We maintain that provider standards and make sure all resources / datasources has its own test.go file
Can you please add individual test.go files..

**Comment by JonWoodlief** [comment]
Screen Shot 2021-03-25 at 9 25 35 AM

Updated with the additional tests

File: ibm/data_source_ibm_cm_catalog.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/**
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_catalog.go:1]: change copyright lines to 

```
// Copyright IBM Corp. 2017, 2021 All Rights Reserved.
// Licensed under the Mozilla Public License v2.0
```
in all files

File: ibm/data_source_ibm_cm_catalog.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+				Required:    true,
+				ForceNew:    true,
+				Description: "ID for catalog",
+			},
+			"rev": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Cloudant revision.",
+			},
+			"label": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_catalog.go:46]: "label","short_description","catalog_icon_url".. All these are response attributes and no where used in request paload..
So, these attributes should be Computed

File: ibm/data_source_ibm_cm_catalog.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		ReadContext: dataSourceIBMCmCatalogRead,
+
+		Schema: map[string]*schema.Schema{
+			"catalog_identifier": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "ID for catalog",
+			},
+			"rev": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_catalog.go:41]: rev declared in schema.. but not set in read method

File: ibm/data_source_ibm_cm_catalog.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+	"github.com/IBM/platform-services-go-sdk/catalogmanagementv1"
+)
+
+func dataSourceIBMCmCatalog() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceIBMCmCatalogRead,
+
+		Schema: map[string]*schema.Schema{
+			"catalog_identifier": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_catalog.go:35]: attribute not set in read method

File: ibm/data_source_ibm_cm_catalog.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	if err = d.Set("label", catalog.Label); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting label: %s", err))
+	}
+	if err = d.Set("short_description", catalog.ShortDescription); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting short_description: %s", err))
+	}
+	if err = d.Set("catalog_icon_url", catalog.CatalogIconURL); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting catalog_icon_url: %s", err))
+	}
+	if err = d.Set("tags", catalog.Tags); err != nil {
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_catalog.go:116]: Use global catalog apis to set "tags"

File: ibm/data_source_ibm_cm_offering.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+			},
+			"offering_support_url": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URL to be displayed in the Consumption UI for getting support on this offering.",
+			},
+			"tags": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				ForceNew:    true,
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_offering.go:91]: computed values cannot be force new..

File: ibm/data_source_ibm_cm_offering.go [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+			},
+			"repo_info": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Repository info for offerings.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"token": &schema.Schema{
+							Type:        schema.TypeString,
+							Optional:    true,
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_offering.go:168]: when parent attribute is computed.. child attributes cannot be optional.. it should computed..

File: ibm/data_source_ibm_cm_offering.go (hunk 6 above)
Comment (kavya498) [reply ibm/data_source_ibm_cm_offering.go:91]: "tags are not set in READ method

File: ibm/data_source_ibm_cm_offering.go [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+				Description: "Catalog identifier.",
+				ForceNew:    true,
+			},
+			"offering_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The id of the catalog containing this offering.",
+				ForceNew:    true,
+			},
+			"rev": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_offering.go:47]: rev is not set in read method

File: ibm/data_source_ibm_cm_offering.go [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+			"disclaimer": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "A disclaimer for this offering.",
+			},
+			"hidden": &schema.Schema{
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Determine if this offering should be displayed in the Consumption UI.",
+			},
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_offering.go:159]: many of the api doc attributes like kind,rating created, updated...are not supported in this datasource.. Is there any specific reason?

File: ibm/data_source_ibm_cm_version.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+		ReadContext: dataSourceIBMCmVersionRead,
+
+		Schema: map[string]*schema.Schema{
+			"version_loc_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Catalog identifier.",
+			},
+			"catalog_identifier": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:41]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Catalog identifier.",
+			},
+			"offering_id": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:46]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Offering identification.",
+			},
+			"tags": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:51]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Tags array.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+			"target_kinds": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:57]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Target kinds.  Current valid values are 'iks', 'roks', 'vcenter', and 'terraform'.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+			"content": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:63]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "byte array representing the content to be imported.  Only supported for OVA images at this time.",
+			},
+			"zipurl": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:68]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URL path to zip location.  If not specified, must provide content in the body of this call.",
+			},
+			"target_version": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:73]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The semver value for this new version, if not found in the zip url package content.",
+			},
+			"repo_type": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:78]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The type of repository containing this version.  Valid values are 'public_git' or 'enterprise_git'.",
+			},
+			"rev": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:83]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "hash of the content.",
+			},
+			"catalog_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Catalog ID.",
+			},
+			"kind_id": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:108]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Content's source URL (e.g git repo).",
+			},
+			"tgz_url": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "File used to on-board this version.",
+			},
+			"single_instance": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:128]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Denotes if single instance can be deployed to a given cluster.",
+			},
+			"image_manifest_url": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:133]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "If set, denotes a url to a YAML file with list of container images used by this version.",
+			},
+			"deprecated": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:138]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "read only field, indicating if this version is deprecated.",
+			},
+			"package_version": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:143]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Version of the package used to create this version.",
+			},
+			"version_locator": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:148]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "A dotted value of `catalogID`.`versionID`.",
+			},
+			"console_url": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:153]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Console URL.",
+			},
+			"long_description": &schema.Schema{
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:158]: not set in read method

File: ibm/data_source_ibm_cm_version.go [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Long description for version.",
+			},
```
Comment (kavya498) [inline ibm/data_source_ibm_cm_version.go:162]: Other response attributes like portal_approval_record, portal_ui_url,repo_info,provider.. are not supported in this datasource.. Any specific reason for that?

File: ibm/resource_ibm_cm_catalog.go [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "CRN associated with the catalog.",
+			},
+			"offerings_url": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "URL path to offerings.",
+			},
+			"created": &schema.Schema{
```
Comment (kavya498) [inline ibm/resource_ibm_cm_catalog.go:81]: not set in read method

File: ibm/resource_ibm_cm_catalog.go [hunk 29]
```diff
@@ -0,0 +1,193 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The date-time this catalog was created.",
+			},
+			"updated": &schema.Schema{
```
Comment (kavya498) [inline ibm/resource_ibm_cm_catalog.go:86]: not set in read method

File: ibm/resource_ibm_cm_catalog.go [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+
+func resourceIBMCmCatalog() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceIBMCmCatalogCreate,
+		Read:     resourceIBMCmCatalogRead,
+		Delete:   resourceIBMCmCatalogDelete,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"rev": &schema.Schema{
```
Comment (kavya498) [inline ibm/resource_ibm_cm_catalog.go:36]: Not set in read method..
why is this attribute only computed?
Can be sent as part of request payload according to API Docs

File: ibm/resource_ibm_cm_catalog_test.go [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+		if err != nil {
+			return err
+		}
+
+		getCatalogOptions := &catalogmanagementv1.GetCatalogOptions{}
+
+		getCatalogOptions.SetCatalogIdentifier(rs.Primary.ID)
+
+		_, _, err = catalogManagementClient.GetCatalog(getCatalogOptions)
+		if err != nil {
```
Comment (kavya498) [inline ibm/resource_ibm_cm_catalog_test.go:80]: return nil if error response is 404

File: website/docs/d/cm_catalog.html.markdown [hunk 32]
```diff
@@ -0,0 +1,45 @@
+---
+layout: "ibm"
+page_title: "IBM : cm_catalog"
+sidebar_current: "docs-ibm-datasource-cm-catalog"
```
Comment (kavya498) [inline website/docs/d/cm_catalog.html.markdown:4]: 1. Add you service name in https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/website/allowed-subcategories.txt
2. remove sidebar_current in all your docs
3. Add sub category at the beginning soomething like..
`subcategory: "Catalog Management`
Refer other docs for reference

**Comment by JonWoodlief** [comment]
@kavya498 

Thanks for the feedback, I've updated the PR

**Comment by JonWoodlief** [comment]
Screen Shot 2021-03-30 at 8 57 20 AM

File: website/allowed-subcategories.txt [hunk 33]
```diff
@@ -21,6 +22,4 @@ Schematics
 Secrets Manager
 Transit Gateway
 VPC infrastructure
-
-
-
+Secrets Manager
```
Comment (kavya498) [inline website/allowed-subcategories.txt:25]: Revert back this Secret manage change

File: website/docs/d/cm_catalog.html.markdown [hunk 34]
```diff
@@ -0,0 +1,45 @@
+---
+layout: "ibm"
+page_title: "IBM : cm_catalog"
+subcategory: "Catalog Management"
```
Comment (kavya498) [inline website/docs/d/cm_catalog.html.markdown:4]: this line should be above  `layout`.. something like
```
subcategory: "Catalog Management"
layout: "ibm"
```
please make this change in all other docs

**Comment by JonWoodlief** [comment]
@kavya498 those changes have been made

**Pull request merged**